csharp/src/SeedLang.Ast/AstStringBuilder.cs

```diff
@@ -56,20 +63,24 @@ internal sealed class AstStringBuilder : AstVisitor {
       return asb.ToString();
     }
 
-    protected internal override void VisitBinaryExpression(BinaryExpression binary) {
+    protected internal override NullResult VisitBinaryExpression(BinaryExpression binary) {
       _out.Append($"({binary.Left} {binary.Op.Symbol()} {binary.Right})");
+      return null;
```

Comment:
The combination of `NullResult VisitFunc(...)` and `return null` is really confusing to me. Here I think "NullResult" intends to imply "no result will be returned", instead of "it may return a null result to indicate something interesting".

It's okay to use generic patterns to cover a function with different return types. But It sounds a bit weird to me to use a generic pattern to cover both the function's return types and the case that the function returns nothing.

How about the following way:

```
class NoResult {
  public const NoResult Nothing=>null;
}

// For the visitor functions that return a value:
NumberValue VisitBinaryExpression(...)

// For the visitor functions that do not return any value:
NoResult VisitBinaryExpression(...; return NoResult.Nothing; )
```

Or, I think we can also consider an OO solution, like:

```
class Result {...}
class NoResult : Result {...}
class BaseValue : Result {...}
class NumberValue : BaseValue {...}

// Then the base class of all visitors can define the abstract methods like:
Result VisitBinaryExpression(...);

// The concrete visitor with no result could define the method as:
NoResult VisitBinaryExpression(...; return NoResult.Nothing;
```

Reply:
The difficulty is void or Void is not allowed to be used as template parameter in C#. But it's allowed in C++. There are some consideration in this design:

1. The results of visit methods are needed, otherwise the visitor need maintain complex states during the AST tree tranversal.
2. The AstVisitor can be used by several clients. For example, the compiler in the interpreter will use it to travers the AST tree to convert it to bytecodes. In this situation, the return value might be a class defined bby the compiler. I don't think there is any relationship between this class and the BaseValue. So I think generic result type can express it well.

So maybe use your first solution now. We can change it later if we have better idea.

Reply:
Agreed.